internal/controllers/operator_controller_test.go

```diff
@@ -32,1028 +30,1234 @@ import (
+	require.Equal(t, operatorsv1alpha1.ReasonInstallationFailed, cond.Reason)
+	require.Equal(t, "unknown bundle mediatype: badmedia+v1", cond.Message)
+
+	verifyInvariants(ctx, t, reconciler.Client, operator)
+	require.NoError(t, cl.DeleteAllOf(ctx, &operatorsv1alpha1.Operator{}))
+	require.NoError(t, cl.DeleteAllOf(ctx, &rukpakv1alpha1.BundleDeployment{}))
+}
 
-			It("should add an invalid spec condition and *not* re-enqueue for reconciliation", func() {
-				By("running reconcile")
-				res, err := reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: opKey})
-				Expect(res).To(Equal(ctrl.Result{}))
-				Expect(err).ToNot(HaveOccurred())
+func TestOperatorInvalidSemverPastRegex(t *testing.T) {
+	cl, err := newClient()
+	require.NoError(t, err)
+	require.NotNil(t, cl)
+	ctx := context.Background()
+	fakeCatalogClient := testutil.NewFakeCatalogClient(testBundleList)
+	reconciler := &controllers.OperatorReconciler{
+		Client:   cl,
+		Scheme:   sch,
+		Resolver: solver.NewDeppySolver(controllers.NewVariableSource(cl, &fakeCatalogClient)),
+	}
 
-				By("fetching updated operator after reconcile")
-				Expect(fakeClient.Get(ctx, opKey, operator)).NotTo(HaveOccurred())
+	t.Log("When an invalid semver is provided that bypasses the regex validation")
+	opKey := types.NamespacedName{Name: fmt.Sprintf("operator-validation-test-%s", rand.String(8))}
 
-				By("Checking the status fields")
-				Expect(operator.Status.ResolvedBundleResource).To(Equal(""))
-				Expect(operator.Status.InstalledBundleResource).To(Equal(""))
+	t.Log("By injecting creating a client with the bad operator CR")
+	pkgName := fmt.Sprintf("exists-%s", rand.String(6))
+	operator := &operatorsv1alpha1.Operator{
+		ObjectMeta: metav1.ObjectMeta{Name: opKey.Name},
+		Spec: operatorsv1alpha1.OperatorSpec{
+			PackageName: pkgName,
+			Version:     "1.2.3-123abc_def", // bad semver that matches the regex on the CR validation
+		},
+	}
 
-				By("checking the expected conditions")
-				cond := apimeta.FindStatusCondition(operator.Status.Conditions, operatorsv1alpha1.TypeResolved)
-				Expect(cond).NotTo(BeNil())
-				Expect(cond.Status).To(Equal(metav1.ConditionUnknown))
-				Expect(cond.Reason).To(Equal(operatorsv1alpha1.ReasonResolutionUnknown))
-				Expect(cond.Message).To(Equal("validation has not been attempted as spec is invalid"))
-				cond = apimeta.FindStatusCondition(operator.Status.Conditions, operatorsv1alpha1.TypeInstalled)
-				Expect(cond).NotTo(BeNil())
-				Expect(cond.Status).To(Equal(metav1.ConditionUnknown))
-				Expect(cond.Reason).To(Equal(operatorsv1alpha1.ReasonInstallationStatusUnknown))
-				Expect(cond.Message).To(Equal("installation has not been attempted as spec is invalid"))
-			})
-		})
-	})
-})
+	// this bypasses client/server-side CR validation and allows us to test the reconciler's validation
+	fakeClient := fake.NewClientBuilder().WithScheme(sch).WithObjects(operator).WithStatusSubresource(operator).Build()
```

Comment:
Not something you need to change here but we might want to queue this up for discussion later, this test case seems a bit wild.

Reply:
I didn't originally write most of these tests! :)

Reply:
Yep! Undersood :)